### PR TITLE
fix: #1076 fetch before navigation

### PR DESCRIPTION
**Pull request by ianliuwk1019**
- [x] Use singleton for ApiServiceFactory and export services instead, then remove usage of Provide/Inject.
- [x] Organize some data fetching code into fetchData.ts.
- [x] Add router beforeEnter() to handle dataFetching before navigating to components.
- [x] Minor naming refactoring.

### CODE REVIEW

File: frontend/src/store/UserAccessRoleState.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import type { FamApplicationUserRoleAssignmentGet } from 'fam-app-acsctl-api';
+import { shallowRef } from 'vue';
+
+export const userRoleAssignments = shallowRef<
+    FamApplicationUserRoleAssignmentGet[]
+>([]);
+
+// --- Setter
+
+export const setUserRoleAssignments = (
```
Comment (MCatherine1994) [inline frontend/src/store/UserAccessRoleState.ts:10]: Hi Ian, I feel normally we set state variables when we want to share the value between different components. But in our case, only the ManagePermission component needs this value, is that better if we pass it as prop as well from router? and still use local variable to handle?

Comment (ianliuwk1019) [reply frontend/src/store/UserAccessRoleState.ts:10]: Thanks Catherine, initially I feel `userRoleAssignments` should be a global state, due to the fact that "**ManagePermission**" seems to be our top level component (dashboard) already. I may have overlook that time.
 
When I look again, the "**userRoleAssignments**" are shared between "**ManagePermission**" and child component "**UserDataTable**" component.

I wasn't aware this "**userRoleAssignments**" was passed to "**UserDataTable**" component as a **props**. Now I feel even "UserDataTable" does not need this props, it should reflect right away what's changed in the state to show the table. 

Question, do you think in the future "**UserDataTable**" is going to be **used by different tabs**, so the props is needed (I kind of think it is going to be stay only specific for showing and sharing "**userRoleAssignment**") ?

Comment (ianliuwk1019) [reply frontend/src/store/UserAccessRoleState.ts]: Hi @MCatherine1994 I have switched this to be as props. Hope I didn't miss anything.

File: frontend/src/components/managePermissions/ManagePermissions.vue [hunk 2]
```diff
@@ -14,78 +14,44 @@ import {
...
-        await getAppUserRoleAssignment();
-    }
+import {
+    deletAndRefreshUserRoleAssignments,
+    fetchUserRoleAssignments,
+} from '@/services/fetchData';
+
+const props = defineProps({
+    userRoleAssignments: {
+        // options fetched from route.
```
Comment (MCatherine1994) [inline frontend/src/components/managePermissions/ManagePermissions.vue:28]: shall we update this comment?

File: frontend/src/services/fetchData.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import { AppActlApiService } from '@/services/ApiServiceFactory';
+import { setApplicationsUserAdministers } from '@/store/ApplicationState';
+import type { FamApplicationUserRoleAssignmentGet } from 'fam-app-acsctl-api';
+
+// --- Fetching data (from backend)
```
Comment (MCatherine1994) [inline frontend/src/services/fetchData.ts:5]: Hi Ian, this file looks good to me, but I wonder if we could have another name? This is like a central place to make service calls including get and delete. I know in Redux these are called actions. Maybe something like appActlApiActions?  I'm really bad at naming ....

File: frontend/src/components/managePermissions/ManagePermissions.vue (hunk 2 above)
Comment (ianliuwk1019) [reply frontend/src/components/managePermissions/ManagePermissions.vue]: This is not necessary, I removed it.

File: frontend/src/services/fetchData.ts (hunk 3 above)
Comment (ianliuwk1019) [reply frontend/src/services/fetchData.ts:5]: I don't use Redux before and don't know what actions does. I think it is for state management.
I don't prefer terms stick to a particular technology stacks unless the terms really speaks to itself, like "store/state".
If other people use different stack then it isn't clear what "actions" do to them and may confuse others.
The fetchData is created just for convenience on taking care of the calls (really, it is just to fetching the data from using the api(s) ) many components could face and does not limit to just one API (unless we like to). I basically just like to fetch data from database (and some minor/general processing I see fit after data is fetched), does not involve state if it does not need to.

**Review by MCatherine1994 — APPROVED** [review]

**Pull request merged**